File: Extra Practice/Self Practice/Sorting/quick_sort.js
function quick_sort(arr, start, end){
    if(start<end){
        let pivot_idx = parition(arr, start, end)
        quick_sort(arr, start, pivot_idx -1);
        quick_sort(arr, pivot_idx+1, end);
    }
}

function parition(arr, start, end){
    let pivot_idx = Math.floor(Math.random() * (end - start ) + start);
    [arr[pivot_idx], arr[end]] = [arr[end], arr[pivot_idx]];
    let pivot_value = arr[end]
    let i=start;
    for(let j=start; j<end; j++){
        if(arr[j] < pivot_value){
            [arr[i], arr[j] ] = [arr[j], arr[i]];
            i++;
        }
    }

    [arr[end], arr[i]] = [arr[i], arr[end]]
    return i;
}


const arr = [9,6,3,1,4,8]
quick_sort(arr, 0, arr.length-1)
console.log(arr)

File: Extra Practice/Self Practice/Sorting/quick_Sort.java
public class quick_Sort {
    public static void swap(int[] arr, int i, int j){
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    public static int parition(int[] arr, int start, int end){
        int pivot_idx = (int)(Math.random() * (end-start))+start;
        int pivot_value = arr[pivot_idx];
        swap(arr, pivot_idx, end);
        int j =start;
        for(int i=start; i<end; i++){
            if(arr[i] < pivot_value){
                swap(arr, i, j);
                j++;
            }
        }
        swap(arr, j, end);
        return j;
    }

    public static int[] sort(int[] arr, int start, int end){
        if(start<end){
            int pivot_idx = parition(arr, start, end); 
            sort(arr, start, pivot_idx-1);
            sort(arr, pivot_idx+1, end);
        }
        return arr;
    }

    public static void main(String[] args) {
        int[] arr = {-1,4,5,6,10,24,2,9};
        arr = sort(arr, 0, arr.length-1);
        for(int num: arr){
            System.out.print(num + " ");
        }
    }
}
